src/wasm/wasm-s-parser.cpp

```diff
@@ -2093,17 +2093,19 @@ Expression* SExpressionWasmBuilder::makeI31Get(Element& s, bool signed_) {
 }
 
 Expression* SExpressionWasmBuilder::makeRefTest(Element& s) {
-  auto ret = allocator.alloc<RefTest>();
-  WASM_UNREACHABLE("TODO (gc): ref.test");
-  ret->finalize();
-  return ret;
+  auto heapType = parseHeapType(*s[1]);
```

Comment:
The text format only has one static heap type even though the binary format has two?

Reply:
I believe so? [MVP doc](https://github.com/WebAssembly/gc/blob/master/proposals/gc/MVP.md#casts) says
```
ref.test $t : [(ref null ht) (rtt n? $t)] -> [i32]
```
for the text, and [binary format doc](https://docs.google.com/document/d/1DklC3qVuOdLHSXB5UXghM_syCh-4cMinQ50ICiXnK3Q/edit?pli=1#) says
```
ref.test | 0xfb40 | <heaptype1> <heaptype2> | [(ref null t1) (rtt n t2)] -> [i32]
```
None of these heap types are actually necessary, so I do hope we can get everyone to agree to remove them before we ship... but maybe there is some baseline compiler angle that I am missing, that makes it better to add them? Anyhow, we can figure that out later I guess.

@jakobkummerow am I misreading the binary format doc perhaps?

Reply:
The "binary format doc" reflects an earlier version of MVP.md, and we're intentionally keeping that stable for now [1]. See comments on that doc for what the V8 prototype currently implements (ref.test and ref.cast with heap type immediates, rtt.sub and br_on_cast without -- it's a bit inconsistent, sorry about that).

I agree that it would be both feasible from an implementation PoV, and desirable from a module size PoV, to drop these type immediates that can easily be inferred by looking at the types of the values on the expression stack.

[1] See the very top where it says "version 2". We're only making backwards-compatible changes at the moment (e.g. Manos added `anyref` support recently), so that toolchains have a non-moving target to work with. That means our prototype does not reflect the latest changes to MVP.js, such as RTTs without depth. At some point we'll define a "version 3" of the prototype spec doc and update the prototype's implementation accordingly; timeline TBD. I'll communicate it early and widely when that happens.

Reply:
I see, thanks @jakobkummerow . I may have been doing the wrong thing then, I've only read the binary format doc when something wasn't clear enough in the MVP. So there may be small differences atm between binaryen and v8/wabt. I'll do a pass eventually to reconcile all that, maybe a good time would be at version 3 of that doc, or when a user is ready to use it, whichever is earlier.